File: merge_viewmodel.py
import re

with open('multi_platform/bizgrow-kmp/shared/src/commonMain/kotlin/com/upstyle/bizgrow/ui/AppViewModel.kt', 'r') as f:
    current_code = f.read()

with open('AppViewModel_ab30789.kt', 'r') as f:
    old_code = f.read()

# Find all functions in old_code
def extract_functions(code):
    funcs = {}
    pattern = re.compile(r'^\s*fun\s+([a-zA-Z0-9_]+)\([^)]*\)(?:\s*:\s*[a-zA-Z0-9_<>]+)?\s*=\s*viewModelScope\.launch\s*\{', re.MULTILINE)
    for match in pattern.finditer(code):
        func_name = match.group(1)
        start = match.end() - 1 # start at {
        brace_count = 0
        end = -1
        in_string = False
        escape = False
        for i in range(start, len(code)):
            c = code[i]
            if escape:
                escape = False
                continue
            if c == '\\':
                escape = True
                continue
            if c == '"':
                in_string = not in_string
                continue
            if not in_string:
                if c == '{':
                    brace_count += 1
                elif c == '}':
                    brace_count -= 1
                    if brace_count == 0:
                        end = i + 1
                        break
        if end != -1:
            funcs[func_name] = code[match.start():end]
    return funcs

old_funcs = extract_functions(old_code)

# Now find all stubs in current_code and replace them
pattern_stub = re.compile(r'^(\s*)fun\s+([a-zA-Z0-9_]+)\([^)]*\)\s*=\s*viewModelScope\.launch\s*\{\s*\}', re.MULTILINE)

def replace_stub(match):
    indent = match.group(1)
    func_name = match.group(2)
    if func_name in old_funcs:
        # replace with old_funcs[func_name]
        body = old_funcs[func_name]
        # adjust indentation slightly if needed, but it's probably fine
        return body
    else:
        print(f"Warning: stub for {func_name} not found in old code")
        return match.group(0)

merged_code = pattern_stub.sub(replace_stub, current_code)

with open('multi_platform/bizgrow-kmp/shared/src/commonMain/kotlin/com/upstyle/bizgrow/ui/AppViewModel.kt', 'w') as f:
    f.write(merged_code)

print("Merge complete")

File: merge_viewmodel.js
const fs = require('fs');

const currentCode = fs.readFileSync('multi_platform/bizgrow-kmp/shared/src/commonMain/kotlin/com/upstyle/bizgrow/ui/AppViewModel.kt', 'utf8');
const oldCode = fs.readFileSync('AppViewModel_ab30789.kt', 'utf8');

function extractFunctions(code) {
    const funcs = {};
    const pattern = /fun\s+([a-zA-Z0-9_]+)\([^)]*\)(?:\s*:\s*[a-zA-Z0-9_<>]+)?\s*=\s*viewModelScope\.launch\s*\{/g;
    let match;
    while ((match = pattern.exec(code)) !== null) {
        const funcName = match[1];
        const start = match.index + match[0].length - 1; // start at {
        let braceCount = 0;
        let end = -1;
        let inString = false;
        let escape = false;
        for (let i = start; i < code.length; i++) {
            const c = code[i];
            if (escape) { escape = false; continue; }
            if (c === '\\\\') { escape = true; continue; }
            if (c === '"') { inString = !inString; continue; }
            if (!inString) {
                if (c === '{') braceCount++;
                else if (c === '}') {
                    braceCount--;
                    if (braceCount === 0) {
                        end = i + 1;
                        break;
                    }
                }
            }
        }
        if (end !== -1) {
            funcs[funcName] = code.substring(match.index, end);
        }
    }
    return funcs;
}

const oldFuncs = extractFunctions(oldCode);

const stubPattern = /^([ \t]*)fun\s+([a-zA-Z0-9_]+)\([^)]*\)(?:\s*:\s*[a-zA-Z0-9_<>]+)?\s*=\s*viewModelScope\.launch\s*\{\s*\}/gm;

let replacedCount = 0;
const mergedCode = currentCode.replace(stubPattern, (match, indent, funcName) => {
    if (oldFuncs[funcName]) {
        replacedCount++;
        // prefix the old body with the correct indent
        let body = oldFuncs[funcName];
        // The oldFuncs body starts with "fun ... = viewModelScope.launch {" without leading spaces.
        // We need to add indent to every line of body.
        body = body.split('\n').map((line, i) => i === 0 ? indent + line : line).join('\n');
        return body;
    } else {
        console.log("Warning: stub for " + funcName + " not found in old code");
        return match;
    }
});

fs.writeFileSync('multi_platform/bizgrow-kmp/shared/src/commonMain/kotlin/com/upstyle/bizgrow/ui/AppViewModel.kt', mergedCode);
console.log("Merge complete. Replaced " + replacedCount + " stubs.");
